lib/adapters/serializers/csv.js

```diff
@@ -5,50 +5,107 @@ var base = require('./base');
 var csv = require('csv-write-stream');
 var errors = require('../../errors');
 var values = require('../../runtime/values');
+var readline = require('readline');
 
 var CSVSerializer = base.extend({
 
+    initialize: function(stream, options) {
+        options = options  || {};
+
+        if (options.append && !options.input) {
+            throw errors.runtimeError('RT-INTERNAL-ERROR', {
+                error: 'input stream must be provided when appending'
+            });
+        }
+
+        this.append = options.append;
+        this.input = options.input;
+        this.base = Promise.resolve();
+    },
+
     write: function(points) {
-        var self = this;
+        if (this.append && !this.readHeaders) {
+            this.readHeaders = true;
+            this.base = this.base
+            .then(() => {
+                return new Promise((resolve, reject) => {
 
-        _.each(points, function(point) {
-            var keys = _.keys(point);
+                    var liner = readline.createInterface({
+                        input: this.input,
+                        terminal: false
+                    });
 
-            _.each(point, function(value, key) {
-                if (values.isArray(value) || values.isObject(value)) {
-                    point[key] = JSON.stringify(value);
-                }
+                    liner.on('line', (line) => {
+                        // there will be a few lines process before the
+                        // readline instance notices the stream has been closed
+                        if (!this.headers) {
+                            this.sendHeaders = false;
+                            this.headers = line.split(',');
+                            this.input.close();
+                            resolve();
```

Comment:
Just figured out this is a bug becuase liner doens't emit 'end' its 'close' (:sadface:)